File: backend/app/api/v1/purchase_materials.py
from datetime import date, timedelta
from typing import Annotated, Literal

from fastapi import APIRouter, Depends, Query, status
from fastapi.responses import Response

from app.api.deps import OrSearch, OrSearch128, OrSearch255, PageNo, PageSize, SortOrder
from app.core.constants import EXPORT_ROW_LIMIT
from app.core.errors import AppError
from app.core.permissions import (
    CurrentUser,
    DbSession,
    IfMatchVersion,
    PurchaseWriter,
    require_roles,
)
from app.domain.enums import PurchasePlanStatus, Role
from app.models import User
from app.schemas import (
    ApiError,
    BatchMovePurchasePlansRequest,
    BatchUpdatePurchasePlansRequest,
    LinkStockMaterialRequest,
    MovePurchasePlanRequest,
    Page,
    PurchaseFilterOptions,
    PurchaseMaterialCreate,
    PurchaseMaterialRead,
    PurchaseMaterialUpdate,
    PurchasePlanExportRequest,
    PurchasePlanResultColumn,
    PurchasePlanResultExportRequest,
    PurchaseRecordRead,
)
from app.services import (
    ai_search_service,
    excel_export_service,
    material_service,
    purchase_request_service,
)
from app.services.common import validate_version

router = APIRouter(prefix="/purchase-materials", tags=["申购计划"])
PlanSearchField = Literal[
    "plan_no",
    "plan_date",
    "material_code",
    "category",
    "name",
    "model_spec",
    "unit_name",
    "planned_qty",
    "usage",
    "subitem_no",
    "remark",
]
LinkWriter = Annotated[
    User,
    Depends(require_roles(Role.SUPER_ADMIN, Role.WAREHOUSE_ADMIN, Role.PURCHASE_ADMIN)),
]
PLAN_RESULT_HEADERS = {
    "plan_no": "计划 ID",
    "plan_date": "需求日期",
    "material_code": "物料编码",
    "category": "类别",
    "urgency": "紧急程度",
    "demand_department": "需求部门",
    "name": "名称",
    "model_spec": "型号规格",
    "planned_qty": "计划数量",
    "unit_name": "计量单位",
    "actual_demand_person": "实际需求人",
    "purchase_responsible": "申购负责人",
    "subitem_no": "子项号",
    "usage": "用途",
}


@router.get("", response_model=Page[PurchaseMaterialRead])
async def list_materials(
    session: DbSession,
    user: CurrentUser,
    page: PageNo = 1,
    page_size: PageSize = 20,
    keyword: OrSearch = None,
    search_field: PlanSearchField | None = None,
    search_value: OrSearch255 = None,
    name: OrSearch128 = None,
    model_spec: OrSearch255 = None,
    actual_demand_person: OrSearch128 = None,
    empty_actual_demand_person: bool = False,
    purchase_responsible: OrSearch128 = None,
    subitem_no: Annotated[str | None, Query(max_length=64)] = None,
    empty_subitem_no: bool = False,
    category: Annotated[str | None, Query(max_length=64)] = None,
    status: Annotated[list[PurchasePlanStatus] | None, Query()] = None,
    coded: bool | None = None,
    moved: bool | None = None,
    ai_expand: bool = False,
    sort_by: PurchasePlanResultColumn | None = None,
    sort_order: SortOrder = "asc",
) -> Page[PurchaseMaterialRead]:
    if user.role != Role.SUPER_ADMIN:
        if status and PurchasePlanStatus.ARCHIVED in status:
            raise AppError(
                "ARCHIVED_PURCHASE_PLAN_FORBIDDEN",
                "仅超级管理员可查询已归档申购计划",
                status_code=403,
            )
        status = status or [PurchasePlanStatus.NORMAL]
    if ai_expand:
        keyword = await ai_search_service.expand_search_value(session, keyword)
        name = await ai_search_service.expand_search_value(session, name)
        if search_field == "name":
            search_value = await ai_search_service.expand_search_value(session, search_value)
    items, total = await material_service.search_purchase_materials(
        session,
        keyword=keyword,
        search_field=search_field,
        search_value=search_value,
        name=name,
        model_spec=model_spec,
        actual_demand_person=actual_demand_person,
        empty_actual_demand_person=empty_actual_demand_person,
        purchase_responsible=purchase_responsible,
        subitem_no=subitem_no,
        empty_subitem_no=empty_subitem_no,
        category=category,
        status=status,
        coded=coded,
        moved=moved,
        page=page,
        page_size=page_size,
        sort_by=sort_by,
        sort_order=sort_order,
    )
    moved_ids = await material_service.purchase_material_ids_moved_to_record(
        session, [item.id for item in items]
    )
    return Page(
        items=[await material_service.purchase_read(session, item, moved_ids) for item in items],
        page=page,
        page_size=page_size,
        total=total,
    )


@router.get("/filter-options", response_model=PurchaseFilterOptions)
async def filter_options(
    session: DbSession,
    user: CurrentUser,
    moved: bool | None = None,
) -> PurchaseFilterOptions:
    (
        actual_demand_persons,
        purchase_responsibles,
        subitem_nos,
        categories,
    ) = await material_service.purchase_filter_options(
        session,
        moved=moved,
        status=None if user.role == Role.SUPER_ADMIN else PurchasePlanStatus.NORMAL,
    )
    return PurchaseFilterOptions(
        actual_demand_persons=actual_demand_persons,
        purchase_responsibles=purchase_responsibles,
        subitem_nos=subitem_nos,
        categories=categories,
    )


@router.post("", response_model=PurchaseMaterialRead, status_code=status.HTTP_201_CREATED)
async def create_material(
    data: PurchaseMaterialCreate, session: DbSession, user: PurchaseWriter
) -> PurchaseMaterialRead:
    item = await material_service.create_purchase_material(session, data)
    return await material_service.purchase_read(session, item)


@router.post("/export-results")
async def export_material_results(
    data: PurchasePlanResultExportRequest,
    session: DbSession,
    user: CurrentUser,
) -> Response:
    export_status = [data.status] if isinstance(data.status, PurchasePlanStatus) else data.status
    if user.role != Role.SUPER_ADMIN:
        if export_status and PurchasePlanStatus.ARCHIVED in export_status:
            raise AppError(
                "ARCHIVED_PURCHASE_PLAN_FORBIDDEN",
                "仅超级管理员可查询已归档申购计划",
                status_code=403,
            )
        export_status = export_status or [PurchasePlanStatus.NORMAL]
    items, total = await material_service.search_purchase_materials(
        session,
        keyword=None,
        search_field=None,
        search_value=None,
        name=data.name,
        model_spec=data.model_spec,
        actual_demand_person=data.actual_demand_person,
        empty_actual_demand_person=data.empty_actual_demand_person,
        purchase_responsible=None,
        subitem_no=data.subitem_no,
        empty_subitem_no=data.empty_subitem_no,
        category=data.category,
        status=export_status,
        coded=None,
        moved=False,
        page=1,
        page_size=EXPORT_ROW_LIMIT + 1,
        sort_by=data.sort_by,
        sort_order=data.sort_order,
    )
    if total > EXPORT_ROW_LIMIT:
        raise AppError(
            "EXPORT_RESULT_LIMIT_EXCEEDED",
            f"查询结果超过 {EXPORT_ROW_LIMIT} 行，请缩小筛选范围后导出",
            status_code=400,
            details={"total": total, "limit": EXPORT_ROW_LIMIT},
        )
    rows = [
        {
            "plan_no": item.plan_no,
            "plan_date": item.plan_date,
            "material_code": item.material_code,
            "category": item.category,
            "urgency": item.urgency,
            "demand_department": item.demand_department,
            "name": item.name,
            "model_spec": item.model_spec,
            "planned_qty": item.planned_qty,
            "unit_name": item.unit_name,
            "actual_demand_person": item.actual_demand_person,
            "purchase_responsible": item.purchase_responsible,
            "subitem_no": item.subitem_no,
            "usage": item.usage,
        }
        for item in items
    ]
    columns = [(key, PLAN_RESULT_HEADERS[key]) for key in data.columns]
    return excel_export_service.excel_response(*excel_export_service.render_result_excel(
      "申购计划导出", columns, rows))


@router.get(
    "/export-uncoded",
    responses={400: {"model": ApiError, "description": "Excel 导出模板缺失或格式错误"}},
)
async def export_uncoded_materials(
    session: DbSession,
    user: CurrentUser,
    keyword: str | None = None,
) -> Response:
    materials = await material_service.purchase_materials_for_export(
        session,
        keyword=keyword,
        coded=False,
        moved=False,
        status=PurchasePlanStatus.NORMAL,
    )
    rows = [
        {
            "serial": index,
            "name": item.name,
            "model_spec": item.model_spec,
            "unit_name": item.unit_name,
            "warranty_managed": "否",
            "asset_category": "非资产",
            "application_reason": "当前无准确编码对应，需要新增编码",
            "department": "HXNI 检修维护部",
            "actual_demand_person": item.actual_demand_person,
        }
        for index, item in enumerate(materials, start=1)
    ]
    return excel_export_service.excel_response(
        *excel_export_service.render_excel("material-code-application.json", rows)
    )


@router.post(
    "/export-purchase-application",
    responses={400: {"model": ApiError, "description": "Excel 导出模板缺失或格式错误"}},
)
async def export_purchase_application(
    data: PurchasePlanExportRequest,
    session: DbSession,
    user: CurrentUser,
) -> Response:
    materials = await material_service.purchase_materials_for_export(
        session,
        material_ids=data.material_ids,
        coded=None,
        moved=False,
        status=None if user.role == Role.SUPER_ADMIN else PurchasePlanStatus.NORMAL,
    )
    material_service.validate_purchase_application_export(materials)
    rows = [
        {
            "material_code": item.material_code,
            "name": item.name,
            "planned_qty": item.planned_qty,
            "actual_demand_person": item.purchase_responsible,
            "demand_department": item.demand_department,
            "required_arrival_date": date.today() + timedelta(days=90),
            "urgency": item.urgency,
            "usage": item.usage,
            "remark": item.remark,
            "subitem_no": item.subitem_no,
        }
        for item in materials
    ]
    return excel_export_service.excel_response(*excel_export_service.render_excel(
      "purchase-application.json", rows))


@router.post("/batch-move-to-record", response_model=list[PurchaseRecordRead])
async def batch_move_to_record(
    data: BatchMovePurchasePlansRequest,
    session: DbSession,
    user: PurchaseWriter,
) -> list[PurchaseRecordRead]:
    lines = await purchase_request_service.batch_move_plans_to_record(session, data)
    return [purchase_request_service.purchase_record_read(line) for line in lines]


@router.patch("/batch", response_model=list[PurchaseMaterialRead])
async def batch_update_materials(
    data: BatchUpdatePurchasePlansRequest,
    session: DbSession,
    user: PurchaseWriter,
) -> list[PurchaseMaterialRead]:
    items = await material_service.batch_update_purchase_materials(session, data)
    moved_ids = await material_service.purchase_material_ids_moved_to_record(
        session, [item.id for item in items]
    )
    return [
        await material_service.purchase_read(session, item, moved_ids) for item in items
    ]


@router.get("/{material_id}", response_model=PurchaseMaterialRead)
async def material_detail(
    material_id: int, session: DbSession, user: CurrentUser
) -> PurchaseMaterialRead:
    item = await material_service.get_purchase_material(session, material_id)
    if item.status == PurchasePlanStatus.ARCHIVED and user.role != Role.SUPER_ADMIN:
        raise AppError(
            "ARCHIVED_PURCHASE_PLAN_FORBIDDEN",
            "仅超级管理员可查询已归档申购计划",
            status_code=403,
        )
    return await material_service.purchase_read(session, item)


@router.patch("/{material_id}", response_model=PurchaseMaterialRead)
async def update_material(
    material_id: int,
    data: PurchaseMaterialUpdate,
    session: DbSession,
    user: PurchaseWriter,
) -> PurchaseMaterialRead:
    item = await material_service.get_purchase_material(session, material_id)
    item = await material_service.update_purchase_material(session, item, data)
    return await material_service.purchase_read(session, item)


@router.delete("/{material_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_material(
    material_id: int,
    session: DbSession,
    user: PurchaseWriter,
    if_match: IfMatchVersion,
) -> None:
    item = await material_service.get_purchase_material(session, material_id)
    await material_service.delete_purchase_material(session, item, if_match)


@router.post("/{material_id}/link-stock-material", response_model=PurchaseMaterialRead)
async def link_stock_material(
    material_id: int,
    data: LinkStockMaterialRequest,
    session: DbSession,
    user: LinkWriter,
) -> PurchaseMaterialRead:
    item = await material_service.get_purchase_material(session, material_id)
    validate_version(data.version, item.version)
    stock = await material_service._validate_stock_link(session, data.stock_material_id)
    item.stock_material_id = data.stock_material_id
    item.stock_material = stock
    item.version += 1
    await session.flush()
    return await material_service.purchase_read(session, item)


@router.post("/{material_id}/move-to-record", response_model=PurchaseRecordRead)
async def move_to_record(
    material_id: int,
    data: MovePurchasePlanRequest,
    session: DbSession,
    user: PurchaseWriter,
) -> PurchaseRecordRead:
    line = await purchase_request_service.move_plan_to_record(session, material_id, data)
    return purchase_request_service.purchase_record_read(line)

File: server/internal/handler/purchase_materials.go
package handler

import (
	"net/http"
	"time"

	"github.com/gin-gonic/gin"

	"github.com/yangrucheng/materials-manager/server/internal/auth"
	"github.com/yangrucheng/materials-manager/server/internal/binding"
	"github.com/yangrucheng/materials-manager/server/internal/domain"
	"github.com/yangrucheng/materials-manager/server/internal/dto"
	"github.com/yangrucheng/materials-manager/server/internal/respond"
	"github.com/yangrucheng/materials-manager/server/internal/serialize"
	"github.com/yangrucheng/materials-manager/server/internal/service"
)

// PurchaseMaterialsHandler 申购计划。
type PurchaseMaterialsHandler struct {
	App *App
}

func NewPurchaseMaterialsHandler(app *App) *PurchaseMaterialsHandler {
	return &PurchaseMaterialsHandler{App: app}
}

func purchaseRead(db *gormDB, item *modelsPurchaseMaterial, moved map[int64]bool) *dto.PurchaseMaterialRead {
	data, appErr := service.LoadPurchaseMaterialReadData(db, item, moved)
	if appErr != nil {
		return nil
	}
	images := make([]dto.FileObjectRead, 0, len(data.Images))
	for i := range data.Images {
		images = append(images, dto.NewFileObjectRead(&data.Images[i]))
	}
	return &dto.PurchaseMaterialRead{
		ID: item.ID, PlanNo: item.PlanNo, PlanDate: serialize.Date(item.PlanDate),
		MaterialCode: item.MaterialCode, Category: item.Category, Urgency: item.Urgency,
		DemandDepartment: item.DemandDepartment, Name: item.Name, ModelSpec: item.ModelSpec,
		UnitName: item.UnitName, ActualDemandPerson: item.ActualDemandPerson,
		PurchaseResponsible: item.PurchaseResponsible,
		PlannedQty:          serialize.DecimalToString(item.PlannedQty.Decimal),
		Usage:               item.Usage, SubitemNo: item.SubitemNo, Remark: item.Remark,
		StockMaterialID: item.StockMaterialID, StockMaterialName: data.StockMaterialName,
		Status: domain.PlanStatusValue[item.Status], MovedToRecord: data.MovedToRecord,
		Images: images, CreatedAt: serialize.UTCZTime(item.CreatedAt),
		UpdatedAt: serialize.UTCZTime(item.UpdatedAt), Version: item.Version,
	}
}

// List GET /purchase-materials
func (h *PurchaseMaterialsHandler) List(c *gin.Context) {
	page, appErr := binding.QueryInt(c, "page", 1, 1, 1<<31-1)
	if appErr != nil {
		respond.Error(c, appErr)
		return
	}
	pageSize, appErr := binding.QueryInt(c, "page_size", 20, 1, 200)
	if appErr != nil {
		respond.Error(c, appErr)
		return
	}
	user := auth.CurrentUser(c)
	isSuper := user != nil && user.Role == domain.RoleSuperAdmin
	var status []string
	for _, s := range c.QueryArray("status") {
		if name, ok := domain.PlanStatusName[s]; ok {
			status = append(status, name)
		}
	}
	if !isSuper {
		if contains(status, domain.PlanArchived) {
			respond.Error(c, archivedForbidden())
			return
		}
		if len(status) == 0 {
			status = []string{domain.PlanNormal}
		}
	}
	coded := parseOptionalBool(c, "coded")
	moved := parseOptionalBool(c, "moved")
	items, total, appErr := service.SearchPurchaseMaterials(h.App.DB, c.Query("keyword"),
		c.Query("search_field"), c.Query("search_value"), c.Query("name"), c.Query("model_spec"),
		c.Query("actual_demand_person"), parseBool(c, "empty_actual_demand_person"),
		c.Query("purchase_responsible"), c.Query("subitem_no"), parseBool(c, "empty_subitem_no"),
		c.Query("category"), status, coded, moved, page, pageSize,
		c.Query("sort_by"), c.Query("sort_order"))
	if appErr != nil {
		respond.Error(c, appErr)
		return
	}
	ids := make([]int64, 0, len(items))
	for i := range items {
		ids = append(ids, items[i].ID)
	}
	movedMap, _ := service.PurchaseMaterialIDsMovedToRecord(h.App.DB, ids)
	reads := make([]dto.PurchaseMaterialRead, 0, len(items))
	for i := range items {
		reads = append(reads, *purchaseRead(h.App.DB, &items[i], movedMap))
	}
	respond.JSON(c, http.StatusOK, dto.Page[dto.PurchaseMaterialRead]{
		Items: reads, Page: page, PageSize: pageSize, Total: total,
	})
}

func archivedForbidden() *appErr {
	return appErrNew("ARCHIVED_PURCHASE_PLAN_FORBIDDEN", "仅超级管理员可查询已归档申购计划", 403, nil)
}

// FilterOptions GET /purchase-materials/filter-options
func (h *PurchaseMaterialsHandler) FilterOptions(c *gin.Context) {
	user := auth.CurrentUser(c)
	isSuper := user != nil && user.Role == domain.RoleSuperAdmin
	var status []string
	if !isSuper {
		status = []string{domain.PlanNormal}
	}
	moved := parseOptionalBool(c, "moved")
	persons, responsibles, subitems, categories, appErr := service.PurchaseFilterOptions(h.App.DB, moved, status)
	if appErr != nil {
		respond.Error(c, appErr)
		return
	}
	respond.JSON(c, http.StatusOK, dto.PurchaseFilterOptions{
		ActualDemandPersons: persons, PurchaseResponsibles: responsibles,
		SubitemNos: subitems, Categories: categories,
	})
}

// Create POST /purchase-materials
func (h *PurchaseMaterialsHandler) Create(c *gin.Context) {
	var req dto.PurchaseMaterialCreate
	if appErr := binding.Body(c, &req); appErr != nil {
		respond.Error(c, appErr)
		return
	}
	item, appErr := createPurchaseMaterial(h.App.DB, &req.PurchaseMaterialBase)
	if appErr != nil {
		respond.Error(c, appErr)
		return
	}
	respond.JSON(c, http.StatusCreated, purchaseRead(h.App.DB, item, nil))
}

func createPurchaseMaterial(db *gormDB, base *dto.PurchaseMaterialBase) (*modelsPurchaseMaterial, *appErr) {
	var planDate *time.Time
	if base.PlanDate != nil {
		if t, err := time.Parse("2006-01-02", *base.PlanDate); err == nil {
			planDate = &t
		}
	}
	status := domain.PlanNormal
	if base.Status != nil {
		status = domain.PlanStatusName[*base.Status]
	}
	return service.CreatePurchaseMaterialFull(db, planDate, derefStr(base.MaterialCode), derefStr(base.Category),
		base.Urgency, base.DemandDepartment, base.Name, base.ModelSpec, base.UnitName,
		base.ActualDemandPerson, base.PurchaseResponsible, base.PlannedQty.Decimal, base.Usage,
		derefStr(base.SubitemNo), derefStr(base.Remark), base.StockMaterialID, status, "", base.ImageIDs)
}

// Detail GET /purchase-materials/{material_id}
func (h *PurchaseMaterialsHandler) Detail(c *gin.Context) {
	materialID, ok := parseIDParam(c, "material_id")
	if !ok {
		return
	}
	item, appErr := service.GetPurchaseMaterial(h.App.DB, materialID)
	if appErr != nil {
		respond.Error(c, appErr)
		return
	}
	user := auth.CurrentUser(c)
	if item.Status == domain.PlanArchived && user != nil && user.Role != domain.RoleSuperAdmin {
		respond.Error(c, archivedForbidden())
		return
	}
	respond.JSON(c, http.StatusOK, purchaseRead(h.App.DB, item, nil))
}

// Update PATCH /purchase-materials/{material_id}
func (h *PurchaseMaterialsHandler) Update(c *gin.Context) {
	materialID, ok := parseIDParam(c, "material_id")
	if !ok {
		return
	}
	var req dto.PurchaseMaterialUpdate
	if appErr := binding.Body(c, &req); appErr != nil {
		respond.Error(c, appErr)
		return
	}
	item, appErr := service.GetPurchaseMaterial(h.App.DB, materialID)
	if appErr != nil {
		respond.Error(c, appErr)
		return
	}
	var planDate *time.Time
	if req.PlanDate != nil {
		if t, err := time.Parse("2006-01-02", *req.PlanDate); err == nil {
			planDate = &t
		}
	}
	var statusPtr *string
	if req.Status != nil {
		s := domain.PlanStatusName[*req.Status]
		statusPtr = &s
	}
	updated, appErr := service.UpdatePurchaseMaterial(h.App.DB, item, planDate,
		derefStr(req.MaterialCode), derefStr(req.Category), req.Urgency, req.DemandDepartment,
		req.Name, req.ModelSpec, req.UnitName, req.ActualDemandPerson, req.PurchaseResponsible,
		req.PlannedQty.Decimal, req.Usage, derefStr(req.SubitemNo), derefStr(req.Remark),
		req.StockMaterialID, statusPtr, req.ImageIDs, req.Version)
	if appErr != nil {
		respond.Error(c, appErr)
		return
	}
	respond.JSON(c, http.StatusOK, purchaseRead(h.App.DB, updated, nil))
}

// Delete DELETE /purchase-materials/{material_id}
func (h *PurchaseMaterialsHandler) Delete(c *gin.Context) {
	materialID, ok := parseIDParam(c, "material_id")
	if !ok {
		return
	}
	item, appErr := service.GetPurchaseMaterial(h.App.DB, materialID)
	if appErr != nil {
		respond.Error(c, appErr)
		return
	}
	if appErr := service.DeletePurchaseMaterial(h.App.DB, item, parseIfMatch(c)); appErr != nil {
		respond.Error(c, appErr)
		return
	}
	c.Status(http.StatusNoContent)
}

// LinkStockMaterial POST /purchase-materials/{material_id}/link-stock-material
func (h *PurchaseMaterialsHandler) LinkStockMaterial(c *gin.Context) {
	materialID, ok := parseIDParam(c, "material_id")
	if !ok {
		return
	}
	var req dto.LinkStockMaterialRequest
	if appErr := binding.Body(c, &req); appErr != nil {
		respond.Error(c, appErr)
		return
	}
	item, appErr := service.GetPurchaseMaterial(h.App.DB, materialID)
	if appErr != nil {
		respond.Error(c, appErr)
		return
	}
	updated, appErr := service.LinkStockMaterial(h.App.DB, item, req.StockMaterialID, req.Version)
	if appErr != nil {
		respond.Error(c, appErr)
		return
	}
	respond.JSON(c, http.StatusOK, purchaseRead(h.App.DB, updated, nil))
}

// MoveToRecord POST /purchase-materials/{material_id}/move-to-record
func (h *PurchaseMaterialsHandler) MoveToRecord(c *gin.Context) {
	materialID, ok := parseIDParam(c, "material_id")
	if !ok {
		return
	}
	var req dto.MovePurchasePlanRequest
	if appErr := binding.Body(c, &req); appErr != nil {
		respond.Error(c, appErr)
		return
	}
	lines, appErr := service.MovePlansToRecord(h.App.DB, []int64{materialID}, nil, "", "", "", "", "", "已申购", "", nil, nil, nil)
	if appErr != nil {
		respond.Error(c, appErr)
		return
	}
	respond.JSON(c, http.StatusOK, recordRead(h.App.DB, &lines[0]))
}

// BatchMoveToRecord POST /purchase-materials/batch-move-to-record
func (h *PurchaseMaterialsHandler) BatchMoveToRecord(c *gin.Context) {
	var req dto.BatchMovePurchasePlansRequest
	if appErr := binding.Body(c, &req); appErr != nil {
		respond.Error(c, appErr)
		return
	}
	ids := make([]int64, 0, len(req.Materials))
	for _, ref := range req.Materials {
		ids = append(ids, ref.ID)
	}
	lines, appErr := service.MovePlansToRecord(h.App.DB, ids, nil, "", "", "", "", "", "已申购", "", nil, nil, nil)
	if appErr != nil {
		respond.Error(c, appErr)
		return
	}
	reads := make([]dto.PurchaseRecordRead, 0, len(lines))
	for i := range lines {
		reads = append(reads, *recordRead(h.App.DB, &lines[i]))
	}
	respond.JSON(c, http.StatusOK, reads)
}

// BatchUpdate PATCH /purchase-materials/batch
func (h *PurchaseMaterialsHandler) BatchUpdate(c *gin.Context) {
	var req dto.BatchUpdatePurchasePlansRequest
	if appErr := binding.Body(c, &req); appErr != nil {
		respond.Error(c, appErr)
		return
	}
	ids := make([]int64, 0, len(req.Materials))
	versions := make([]int, 0, len(req.Materials))
	for _, ref := range req.Materials {
		ids = append(ids, ref.ID)
		versions = append(versions, ref.Version)
	}
	var planDate *time.Time
	if req.PlanDate != nil {
		if t, err := time.Parse("2006-01-02", *req.PlanDate); err == nil {
			planDate = &t
		}
	}
	items, appErr := service.BatchUpdatePurchaseMaterials(h.App.DB, ids, versions, planDate,
		req.Category, req.Urgency, req.DemandDepartment, req.ActualDemandPerson,
		req.PurchaseResponsible, req.SubitemNo, req.Usage, req.Status)
	if appErr != nil {
		respond.Error(c, appErr)
		return
	}
	movedMap, _ := service.PurchaseMaterialIDsMovedToRecord(h.App.DB, ids)
	reads := make([]dto.PurchaseMaterialRead, 0, len(items))
	for i := range items {
		reads = append(reads, *purchaseRead(h.App.DB, &items[i], movedMap))
	}
	respond.JSON(c, http.StatusOK, reads)
}

// RegisterPurchaseMaterials 注册 /purchase-materials 路由。
func RegisterPurchaseMaterials(r *gin.RouterGroup, app *App) {
	h := NewPurchaseMaterialsHandler(app)
	group := r.Group("/purchase-materials", auth.AuthManagement(app.Cfg, app.DB))
	group.GET("", h.List)
	group.GET("/filter-options", h.FilterOptions)
	group.GET("/:material_id", h.Detail)
	write := group.Group("", auth.PurchaseWriter())
	write.POST("", h.Create)
	write.PATCH("/:material_id", h.Update)
	write.DELETE("/:material_id", h.Delete)
	write.POST("/batch", h.BatchUpdate)
	write.POST("/batch-move-to-record", h.BatchMoveToRecord)
	write.POST("/:material_id/move-to-record", h.MoveToRecord)
	linkWriter := auth.RequireRoles("SUPER_ADMIN", "WAREHOUSE_ADMIN", "PURCHASE_ADMIN")
	write.POST("/:material_id/link-stock-material", linkWriter, h.LinkStockMaterial)
}
